smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/TypeScriptSettings.java

```diff
@@ -407,6 +438,44 @@ public SymbolProvider createSymbolProvider(Model model, TypeScriptSettings setti
         }
     }
 
+    /**
+     * An enum indicating the compatibility mode in regards to member nullability.
+     */
+    public enum MemberNullabilityCompatibilityMode {
+        /**
+         * This is the current behavior and it will be the default. When set,
+         * it allows a member that has the {@link RequiredTrait} applied to be "undefined".
+         * By doing so it can still be considered a backwards compatible change fo
+         * Smithy services even when the required constraint is dropped from a member .
+         */
+        RELAXED("relaxed"),
+
+        /**
+         * This will dissallow members marked as {@link RequiredTrait} to be "undefined".
+         */
```

Comment:
I think there should be some warning language here too calling out the risks of using this mode. The risks are that when a server drops `@required` from an output shape (and replaces it with `@default`), if the server does not always serialize a value, customer code consuming the client and trying to access this member, may get a NullPointerException. Smithy [spec](https://smithy.io/2.0/spec/type-refinement-traits.html#default-value-serialization) says
> Authoritative model consumers like servers SHOULD always serialize default values to remove any ambiguity about the value of the most up to default value.

So one should use this mode on the client, only if the server is following this.